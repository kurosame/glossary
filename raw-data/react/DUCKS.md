## category

react

## titles

Ducks
Redux

## description

Redux のデザインパターン  
Action や Reducer を modules にパッケージする

vuejs-boilerplate でも使ってる気がする  
1 ファイルにはしてないけど、actions, getters, mutations, state を import して modules でパッケージして export してる  
そして Store 作るときに modules をセットしている
