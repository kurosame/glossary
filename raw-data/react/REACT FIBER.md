## category

react

## titles

React Fiber
react-fiber

## description

Begin, Complete, Commit のフェーズがある

### Begin, Complete

Component 単位で実行される  
componentWillMount, componentWillReceiveProps, shouldComponentUpdate, componentWillUpdate  
非同期レンダリングでは中断・再開により、複数回ライフサイクル関数が呼ばれる可能性がある

### Commit

全ての Fiber が完了した時に実行される  
DOM 更新などの副作用が発生する処理はここでまとめて反映される  
componentDidMount, componentDidUpdate, componentWillUnmount

### Time Slicing

Fiber により画面のレンダリングが非同期になると、Fiber 単位で処理の優先度を付けることができる  
例えばユーザ操作の優先度を高くすることで、画面がレンダリング中でもユーザ操作を優先して処理することが可能となる

### Suspense

非同期通信をしている間、画面のレンダリングを中断できる  
非同期通信が完了した後に、画面をレンダリングすることが可能  
render に Promise を投げて、その Promise が完了した後に render するみたいな？
