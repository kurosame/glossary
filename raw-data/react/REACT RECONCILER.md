## category

react

## titles

React Reconciler
react-reconciler

## description

VDOM と呼ばれている部分

1. ReactElement のツリーを構築
1. Component をマウント、アンマウント、更新
1. ライフサイクル関数を呼出す
1. 現在の DOM との差分を計算する
1. renderer に差分を伝える
