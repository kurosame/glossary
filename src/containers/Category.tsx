import { AppBar, Tab, Tabs } from '@material-ui/core'
import React from 'react'
import { useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import { States } from '@/modules/states'

interface Props {
  location: { pathname: string }
}

const tabItems: Array<{ label: string; to: string }> = [
  { label: 'Summary', to: '/' },
  { label: 'Apache', to: '/apache' },
  { label: 'Architecture', to: '/architecture' },
  { label: 'AWS', to: '/aws' },
  { label: 'Browser', to: '/browser' },
  { label: 'CI', to: '/ci' },
  { label: 'Compiler', to: '/compiler' },
  { label: 'CSS', to: '/css' },
  { label: 'Design', to: '/design' },
  { label: 'Docker', to: '/docker' },
  { label: 'Firebase', to: '/firebase' },
  { label: 'Go', to: '/go' },
  { label: 'Hardware', to: '/hardware' },
  { label: 'HTML', to: '/html' },
  { label: 'JS', to: '/js' },
  { label: 'Language', to: '/language' },
  { label: 'ML', to: '/ml' },
  { label: 'Mobile', to: '/mobile' },
  { label: 'Module Bundler', to: '/module-bundler' },
  { label: 'Network', to: '/network' },
  { label: 'Package Manager', to: '/package-manager' },
  { label: 'Programming', to: '/programming' },
  { label: 'Python', to: '/python' },
  { label: 'React', to: '/react' },
  { label: 'Scala', to: '/scala' },
  { label: 'Server', to: '/server' },
  { label: 'Swift', to: '/swift' },
  { label: 'Vue.js', to: '/vue' },
  { label: 'Words', to: '/words' },
  { label: 'Workflow Engine', to: '/workflow-engine' }
]

const Category: React.FC<Props> = p => {
  const isLogin = useSelector<States, boolean>(s => s.login.isLogin)

  if (!isLogin) return null
  return (
    <AppBar position="static">
      <Tabs
        value={tabItems.findIndex(o => o.to === p.location.pathname)}
        variant="scrollable"
        scrollButtons="off"
      >
        {tabItems.map(t => {
          const StaticLink = React.forwardRef((pp, ref: React.Ref<Link>) => (
            <Link ref={ref} to={t.to} {...pp} />
          ))
          StaticLink.displayName = 'StaticLink'
          return <Tab key={t.label} label={t.label} component={StaticLink} />
        })}
      </Tabs>
    </AppBar>
  )
}

export default Category
