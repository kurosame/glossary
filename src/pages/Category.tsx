import { AppBar, Tab, Tabs } from '@material-ui/core'
import React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import { States } from '@/modules/states'
import { LoginState } from '@/modules/login'

interface Props {
  state: { login: LoginState }
  location: { pathname: string }
}

const tabItems: Array<{ label: string; to: string }> = [
  { label: 'Summary', to: '/' },
  { label: 'Apache', to: '/apache' },
  { label: 'Architecture', to: '/architecture' },
  { label: 'AWS', to: '/aws' },
  { label: 'Browser', to: '/browser' },
  { label: 'CI', to: '/ci' },
  { label: 'Compiler', to: '/compiler' },
  { label: 'CSS', to: '/css' },
  { label: 'Design', to: '/design' },
  { label: 'Docker', to: '/docker' },
  { label: 'Firebase', to: '/firebase' },
  { label: 'Go', to: '/go' },
  { label: 'Hardware', to: '/hardware' },
  { label: 'HTML', to: '/html' },
  { label: 'JS', to: '/js' },
  { label: 'Language', to: '/language' },
  { label: 'ML', to: '/ml' },
  { label: 'Mobile', to: '/mobile' },
  { label: 'Module Bundler', to: '/module-bundler' },
  { label: 'Network', to: '/network' },
  { label: 'Package Manager', to: '/package-manager' },
  { label: 'Programming', to: '/programming' },
  { label: 'Python', to: '/python' },
  { label: 'React', to: '/react' },
  { label: 'Scala', to: '/scala' },
  { label: 'Server', to: '/server' },
  { label: 'Swift', to: '/swift' },
  { label: 'Vue.js', to: '/vue' },
  { label: 'Words', to: '/words' },
  { label: 'Workflow Engine', to: '/workflow-engine' }
]

const Category = (props: Props): JSX.Element | null =>
  props.state.login.isLogin ? (
    <AppBar position="static">
      <Tabs
        value={tabItems.findIndex(o => o.to === props.location.pathname)}
        scrollable={true}
        scrollButtons="off"
        data-test="category-tabs"
      >
        {tabItems.map(t => (
          <Tab
            key={t.label}
            label={t.label}
            component={(p: {}): JSX.Element => <Link to={t.to} {...p} />}
          />
        ))}
      </Tabs>
    </AppBar>
  ) : null

export default connect((states: States) => ({
  state: { login: states.login }
}))(Category)
