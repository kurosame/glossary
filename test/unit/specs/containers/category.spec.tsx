import Category from '@/containers/Category'
import type { States } from '@/modules/states'
import { cleanup, render, RenderResult } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import { MemoryRouter as Router } from 'react-router-dom'
import configureStore from 'redux-mock-store'

let wrapper: (isLogin: boolean) => RenderResult
beforeEach(() => {
  wrapper = (isLogin): RenderResult => {
    const store = configureStore<Pick<States, 'login'>>()({
      login: { isLogin }
    })
    return render(
      <Provider store={store}>
        <Router initialEntries={['/']}>
          <Category location={{ pathname: '/react' }} />
        </Router>
      </Provider>
    )
  }
})
afterEach(cleanup)

test('Render DOM if `isLogin` is true', () => {
  expect(wrapper(true).container.querySelector('div')).not.toBeNull()
})

test('Not render DOM if `isLogin` is false', () => {
  expect(wrapper(false).container.querySelector('div')).toBeNull()
})

test('Match the snapshot', () => {
  expect(wrapper(true).asFragment()).toMatchSnapshot()
})
