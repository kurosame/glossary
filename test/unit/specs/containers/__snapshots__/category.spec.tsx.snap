// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match the snapshot 1`] = `
<DocumentFragment>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
  >
    <div
      class="MuiTabs-root"
    >
      <div
        class="MuiTabs-scrollable"
        style="width: 99px; height: 99px; position: absolute; top: -9999px; overflow: scroll;"
      />
      <div
        class="MuiTabs-scroller MuiTabs-scrollable"
        style="margin-bottom: 0px;"
      >
        <div
          class="MuiTabs-flexContainer"
          role="tablist"
        >
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              Summary
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/apache"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              Apache
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/architecture"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              Architecture
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/aws"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              AWS
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/browser"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              Browser
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/build"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              Build
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/ci"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              CI
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/compiler"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              Compiler
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/css"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              CSS
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/db"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              DB
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/design"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              Design
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/docker"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              Docker
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/firebase"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              Firebase
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/go"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              Go
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/hardware"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              Hardware
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/html"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              HTML
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/js"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              JS
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/language"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              Language
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/ml"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              ML
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/mobile"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              Mobile
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/network"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              Network
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/package-manager"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              Package Manager
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/programming"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              Programming
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/python"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              Python
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="true"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit Mui-selected"
            href="/react"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              React
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/scala"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              Scala
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/server"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              Server
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/swift"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              Swift
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/vue"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              Vue.js
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/words"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              Words
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
            href="/workflow-engine"
            role="tab"
            tabindex="0"
          >
            <span
              class="MuiTab-wrapper"
            >
              Workflow Engine
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
        </div>
        <span
          class="PrivateTabIndicator-root-141 PrivateTabIndicator-colorSecondary-143 MuiTabs-indicator"
          style="left: 0px; width: 0px;"
        />
      </div>
    </div>
  </header>
</DocumentFragment>
`;
